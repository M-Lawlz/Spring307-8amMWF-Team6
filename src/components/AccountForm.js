import App from "firebase/app";
import "firebase/auth";
import "firebase/firestore";
import { Button, Dialog, DialogContent, DialogTitle } from "@material-ui/core";
import { Form } from "react-bootstrap";
import React from "react";

export default class AccountForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      formType: props.formType ? props.formType : "",
      input: "",
      inputType: "",
      open: props.isFormShowing ? true : false,
      user: null,
    };
  }

  async componentDidMount() {
    await App.auth().onAuthStateChanged((user) => {
      if (user) {
        this.setState({ user: user });
      }
    });
    switch (this.state.formType) {
      case "firstName":
        this.setState({ inputType: "text", printType: "First Name" });
        break;
      case "lastName":
        this.setState({ inputType: "text", printType: "Last Name" });
        break;
      case "email":
        this.setState({ inputType: "email", printType: "Email" });
        break;
      case "password":
        this.setState({ inputType: "password", printType: "Password" });
        break;
      default:
    }
  }

  attemptUpdate = () => {
    const userDoc = App.firestore()
      .collection("users")
      .doc(this.state.user.email);
    if (this.state.formType === "firstName") {
      userDoc
        .update({ firstName: this.state.input })
        .then(this.closeForm)
        .catch(this.handleError);
    } else if (this.state.formType === "lastName") {
      userDoc
        .update({ lastName: this.state.input })
        .then(this.closeForm)
        .catch(this.handleError);
    } else if (this.state.formType === "email") {
      this.state.user
        .updateEmail(this.state.input)
        .then(this.closeForm)
        .catch(this.handleError);
    } else if (this.state.formType === "password") {
      this.state.user
        .updatePassword(this.state.input)
        .then(this.closeForm)
        .catch(this.handleError);
    }
  };

  cancelForm = () => {
    this.setState({ open: false });
    this.props.isFormShowing(false);
  };

  closeForm = () => {
    this.setState({ open: false });
    this.props.isFormShowing(false);
    alert(`Your account information has been updated.`);
  };

  handleError = (error) => {
    alert("Error: " + error.message);
  };

  handleInput = (input) => {
    this.setState({ input: input.target.value });
  };

  submitForm = (event) => {
    event.preventDefault();
    this.attemptUpdate();
  };

  render() {
    if (this.state.open) {
      return (
        <Dialog
          autoDetectWindowHeight={true}
          autoScrollBodyContent={true}
          contentStyle={{ height: 200, width: 300 }}
          modal={true}
          open={this.state.open}
        >
          <DialogTitle style={{ textAlign: "center" }}>
            {`Change ${this.state.printType}`}
          </DialogTitle>
          <DialogContent>
            <Form onSubmit={this.submitForm}>
              <Form.Text>{"Please enter your new information."}</Form.Text>
              <br />
              <br />
              <Form.Group controlId={"formField"}>
                <Form.Control
                  name={"formField"}
                  onInput={this.handleInput}
                  placeholder={this.state.printType}
                  required
                  type={this.state.inputType}
                  value={this.state.input}
                />
              </Form.Group>
              <br />
              <Button color={"primary"} onClick={this.cancelForm}>
                Cancel
              </Button>
              <Button color={"primary"} type={"submit"}>
                {`Change ${this.state.printType}`}
              </Button>
            </Form>
          </DialogContent>
        </Dialog>
      );
    } else {
      return null;
    }
  }
}
