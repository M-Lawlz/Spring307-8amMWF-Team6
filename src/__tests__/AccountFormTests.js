import AccountForm from "../components/AccountForm";
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

window.alert = jest.fn();

test("Mounts firstName formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"firstName"} />);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("firstName");
    expect(accountForm.state("inputType")).toBe("text");
    expect(accountForm.state("printType")).toBe("First Name");
  }, 0);
});

test("Mounts lastName formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"lastName"} />);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("lastName");
    expect(accountForm.state("inputType")).toBe("text");
    expect(accountForm.state("printType")).toBe("Last Name");
  }, 0);
});

test("Mounts email formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"email"} />);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("email");
    expect(accountForm.state("inputType")).toBe("email");
    expect(accountForm.state("printType")).toBe("Email");
  }, 0);
});

test("Mounts password formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"password"} />);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("password");
    expect(accountForm.state("inputType")).toBe("password");
    expect(accountForm.state("printType")).toBe("Password");
  }, 0);
});

test("Mounts profilePicture formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"profilePicture"} />);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("profilePicture");
    expect(accountForm.state("inputType")).toBe("file");
    expect(accountForm.state("printType")).toBe("Profile Picture");
  })
});

test("Mounts coverPhoto formType correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"coverPhoto"}/>);
  setTimeout(() => {
    accountForm.update();
    expect(accountForm.state("formType")).toBe("coverPhoto");
    expect(accountForm.state("inputType")).toBe("file");
    expect(accountForm.state("printType")).toBe("Cover Photo");
  })
});

test("Attempts to update an email form correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"email"} />);
  const updateEmailSpy = jest.fn(() => Promise.resolve());
  accountForm.setState({
    input: "test",
    user: { email: "test@test.com", updateEmail: updateEmailSpy },
  });
  const instance = accountForm.instance();
  instance.attemptUpdate();
  expect(updateEmailSpy).toBeCalledWith("test");
});

test("Attempts to update a password form correctly.", () => {
  const accountForm = shallow(<AccountForm formType={"password"} />);
  const updatePasswordSpy = jest.fn(() => Promise.resolve());
  accountForm.setState({
    input: "test",
    user: { email: "test@test.com", updatePassword: updatePasswordSpy },
  });
  const instance = accountForm.instance();
  instance.attemptUpdate();
  expect(updatePasswordSpy).toBeCalledWith("test");
});

test("Cancels the form correctly.", () => {
  const isFormShowingSpy = jest.fn();
  const accountForm = shallow(<AccountForm isFormShowing={isFormShowingSpy} />);
  const instance = accountForm.instance();
  instance.cancelForm();
  setTimeout(() => {
    expect(accountForm.state("open")).toBe(false);
    expect(isFormShowingSpy).toBeCalledWith(false);
  }, 0);
});

test("Closes the form correctly.", () => {
  jest.spyOn(window, "alert");
  const isFormShowingSpy = jest.fn();
  const accountForm = shallow(<AccountForm isFormShowing={isFormShowingSpy} />);
  const instance = accountForm.instance();
  instance.closeForm();
  setTimeout(() => {
    expect(accountForm.state("open")).toBe(false);
    expect(isFormShowingSpy).toBeCalledWith(false);
    expect(window.alert).toBeCalled();
  }, 0);
});

test("Handles errors correctly.", () => {
  jest.spyOn(window, "alert");
  const accountForm = shallow(<AccountForm />);
  const instance = accountForm.instance();
  instance.handleError({ message: "Test error." });
  expect(window.alert).toBeCalled();
});

test("Handles input correctly.", () => {
  const accountForm = shallow(<AccountForm />);
  const instance = accountForm.instance();
  instance.handleInput({ target: { value: "test" } });
  setTimeout(() => {
    expect(accountForm.state("input")).toBe("test");
  }, 0);
});

test("Submits form correctly.", () => {
  const event = new Event("test");
  jest.spyOn(event, "preventDefault");
  const accountForm = shallow(<AccountForm />);
  accountForm.setState({ user: { email: "test@test.com" } });
  const instance = accountForm.instance();
  jest.spyOn(instance, "attemptUpdate");
  instance.submitForm(event);
  expect(event.preventDefault).toBeCalled();
  expect(instance.attemptUpdate).toBeCalled();
});

test("Renders successfully.", () => {
  const accountForm = shallow(<AccountForm />);
  expect(toJson(accountForm)).toMatchSnapshot();
});
