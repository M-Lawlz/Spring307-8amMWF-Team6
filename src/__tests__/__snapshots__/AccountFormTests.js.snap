// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm renders successfully 1`] = `null`;
